source/common/http/header_map_impl.cc

```diff
@@ -578,7 +602,42 @@ HeaderMap::Lookup HeaderMapImpl::lookup(const LowerCaseString& key,
 void HeaderMapImpl::clear() {
   inline_headers_.clear();
   headers_.clear();
-  cached_byte_size_ = 0;
+}
+
+bool HeaderMapImpl::HeaderList::maybeMakeMap() const {
+  if (lazy_map_.empty()) {
+    if (headers_.size() < MapSizeThreshold) {
+      return false;
+    }
+    for (auto node = headers_.begin(); node != headers_.end(); ++node) {
+      absl::string_view key = node->key().getStringView();
+#if HEADER_MAP_USE_FLAT_HASH_MAP
+      HeaderNodeVector& v = lazy_map_[key];
+      v.push_back(node);
+#endif
+#if HEADER_MAP_USE_MULTI_MAP
+      lazy_map_.insert(std::make_pair(key, node));
+#endif
+    }
+  }
+  return true;
+}
+
+HeaderMapImpl::HeaderLazyMap::iterator
+HeaderMapImpl::HeaderList::find(absl::string_view key) const {
+  if (lazy_map_.empty()) {
```

Comment:
Should this also use maybeMakeMap() to be consistent and also share the construction code?

Reply:
Yeah this could use a refactor; I'll clean that up.